### DISCUSSION TITLE
Support for a 'back' button?

### DISCUSSION THREAD

**Discussion by IrregularShed** [probonopd/MiniDexed#172] [Ideas]
Hi everyone

I've been playing with this for a few days and love it (even if my handy AKAI MPK Play can't control it properly - I think it confuses things by not being a proper MIDI device when it gets powered on by USB, but that's something for another day).

I know the idea is to keep the circuitry required as simple as possible, and I've seen some discussions/issues about clicking/double-clicking confusion and tweaks. My idea is to add support for an optional 'back' button that could be enabled in the minidexed.ini, so you could do single clicks with the rotary encoder switch to select settings in the user interface and use a different button to go back up a level. I started trying to work out how to code this but sadly I just don't have the time at the moment to learn enough about Circle to try and tackle it myself.

What do you think?

**Comment by eLuSiVeMiTe** [comment]
Up down L and R would also work well. 


DOWN to open. Up to back out. L&R to scroll. Wheel to enter value  hold encoder button and L/R to skip op/TG


But I spend a lot of time programming on an MPC so that's my default navigation.

**Reply by IrregularShed** [reply]
I'm trying to imagine that and struggling! I assume you're using a classic MPC because on my MPC One it's all touchscreen 😀

**Comment by probonopd** [comment]
Maybe we can trigger equivalents of the click and double-click events when certain MIDI messages are receievd, so you could use any buttons on MIDI controllers for this purpose.

**Reply by IrregularShed** [reply]
That's a good idea as long as you have buttons to spare! I'm using a Korg nanoKEY2 at the moment, seeing as the AKAI isn't working, and that's not got anything that I could use 😕

**Comment by JohnConnor84** [comment]
I was looking a 4x4 membrane keypad...it connects direct to the GPIO ports and it looks very 80s. Do you guys think it can be configured as 4 navigation arrows, back button and the rest assigned to other basic functions?  My idea would be the membrane keypad at the left side of the LCD and the rotary encoder at the right, so you could easily navigate through.

**Reply by probonopd** [reply]
It's software, so everything is possible. How much work would it be to implement it? I don't know.

**Comment by dcoredump** [comment]
I am a big fan of creating a touch UI for a external HDMI screen (#62). If you don't need a programming interface (e.g. on stage) you don't need to add a screen, but in your studio you have the full access to all parameters.

Circle supports LVGL and there are free editors for creating a LVGL UI available (also commercial like [squareline](https://squareline.io) which looks very promissing, but maybe supports not enough widgets in the free version).

Regards, Holger

**Reply by probonopd** [reply]
Wouldn't you run Dexed on the PC in the studio anyway? And then sysex voices over to MiniDexed?
Or use a phone/tablet app as a "programmer"?

**Reply by dcoredump** [reply]
In fact: no. I am personaly not a friend of "OS based software synths". I have a UI24 as mixing snd recording device, so I really prefer a haptic or UI interface per synth. I really prefer to have the same hardware in studio and on stage.

My concern is that an update or changing any setting on the computer will mess up my performance on stage or create stress in advance.

**Reply by probonopd** [reply]
Makes sense. Well, I assume you are familiar with @positionhigh https://codeberg.org/positionhigh/MicroDexed-touch? Is that roughly what you would have in mind? Probably someone with suitable hardware and enough dedication and time could port it to MiniDexed. Personally I lack the required hardware  and think of MiniDexed more of a headless device operated by other MIDI devices, but surely I wouldn't stop anyone fro such an undertaking.

**Reply by dcoredump** [reply]
> Makes sense. Well, I assume you are familiar with @positionhigh https://codeberg.org/positionhigh/MicroDexed-touch? Is that roughly what you would have in mind? Probably someone with suitable hardware and enough dedication and time could port it to MiniDexed. Personally I lack the required hardware  and think of MiniDexed more of a headless device operated by other MIDI devices, but surely I wouldn't stop anyone fro such an undertaking.

Sure, Mark and I were working on MicroDexed together since a year.

For me a programmer like [DT-7](https://www.dtronics.nl/dt-7) would be very cool. This can be done with a HDMI display which can be connected only when needed.

I am not a friend of tablets and editor software due to bad experiences: in ten or twenty years they may not work anymore. You must know that I used my old DX7II until 2020 on stage... Now it's to heavy for me and that's the reason why I created MicroDexed.

**Reply by probonopd** [reply]
> I am not a friend of tablets and editor software due to bad experiences: in ten or twenty years they may not work anymore.

Exactly my line of thinking. That tablet stuff tends to break after a couple of years because everything is constantly being upgraded. Until it doesn't run on your device anymore.

**Comment by craigyjp** [comment]
An editor I built for the Futur3Soundz XFM2 synth. I'm sure with a little work it can be used to edit the minidexed

![IMG_01052022_222445_(1350_x_1080_pixel)](https://user-images.githubusercontent.com/51026305/166161650-cc178b45-b086-4c67-b8e8-61029c26a2a0.jpg)

**Reply by probonopd** [reply]
Mind blown!

**Reply by positionhigh** [reply]
> An editor I built for the Futur3Soundz XFM2 synth. I'm sure with a little work it can be used to edit the minidexed

This really is impressive. At microdexed, our live FM Sound modulation currently can only do some quick attack/release modifications :_) But i think we could fit something like this on the 320x240 display with touch.. Let me know if that is of any interesting to you.

![](https://codeberg.org/positionhigh/MicroDexed-touch/raw/branch/main/addon/microDexedRemoteConsole/Screenshots/microdexed-002496.png)

Here is a quick video how that is going:
https://youtu.be/4rvezoD7iu8

**Reply by JohnConnor84** [reply]
Hey guys I bought a pack of 5 membrane 4x4 keypads with 8 GPIO pins. Would any of you be (within Germany) interested to get a free spare one and test it? Just PM me and I send you one for free by Post.

**Reply by probonopd** [reply]
Thanks @JohnConnor84. __If__ someone implements the code for it __then__ I'd be happy to receive one for testing. But I fear I won't be able to write the code for it, so it's not worth sending me one unless we find someone who volunteers to write the code.

**Comment by stevelittlefish** [comment]
I too would like a separate button to replace double clicking.  In fact the fiddliness of having to double click the encoder is the only thing making me not want to build one of these!

I'm not very experienced with embedded programming on this scale - this is a fair bit more complicated than things I've worked on.  Looking at the code though, it seems like there's a very simple event system that could be used.

In userinterface.cpp there is this:

```void CUserInterface::EncoderEventHandler (CKY040::TEvent Event)
{
	switch (Event)
	{
	// SNIP
	case CKY040::EventSwitchDoubleClick:
		m_Menu.EventHandler (CUIMenu::MenuEventSelect);
		break;
	// SNIP
	}
}
```

Would it be possible to monitor / poll one of the other GPIO pins and call this event handler whenever the button is pressed?

Ideally I think I would use a non-clickable encoder, and wire up 2 buttons.  One of them would be wired as the encoder click, and the second would be set to trigger the double click event.

I'm more than happy to help out if someone can nudge me in the right direction!

**Reply by probonopd** [reply]
Hello @stevelittlefish. Have a look at https://github.com/rsta2/circle/issues/292 on how to generate interrupts from GPIO lines.

**Reply by stevelittlefish** [reply]
I've ordered the parts - I'll have a go at this next week and report back!

**Reply by diyelectromusic** [reply]
Did you get a chance to look at this?  Alternatively, is anyone looking at button navigation more generally?  I might have a poke about if no-one else is, but is there a strategy for handling GPIO already?  I see all the encoder stuff is wrapped up in the encoder handling in circle itself linked to userinterface.cpp

**Reply by diyelectromusic** [reply]
Well I've had a go (and managed to get in my usual mess with GitHub... hey ho...).  Still if someone would like to see my very, very draft attempt, have a look at: https://github.com/diyelectromusic/MiniDexed/tree/buttongpio

It is only using very crude debouncing at the moment though, so that needs looking at, at the very least.  And I wasn't quite sure how often the update function would be called.  This is just using polling, not interrupts, but it isn't particularly clear how to do GPIO input in circle!

Once again, my C++ is highly likely to be a bit dodgy, so if someone would like to tell me how I should have done it properly, please do! :)

Kevin

**Reply by stevelittlefish** [reply]
I had to wait for the stuff to arrive and then I got really busy with work.

Here are my (opinionated) thoughts:

- It would be better to use the word "button" instead of BTN as this seems to be more consistent with the naming in the rest of the codebase
- If the rest of the UI uses interrupts we should really figure out how to implement them.  I couldn't see how to do it either!  But I'm sure there is a way...
- Calling the buttons "up", "down", "left" and "right" is very specific.  I would personally just like a "select" and "back" button which wouldn't fit with your scheme.  It would be better to use generic names like "menu_next" and "menu_prev" than "left" and "right" because you can lay the buttons out physically however you like

This project is next on my list as soon as I get some free time - I'm more than happy to help improve some of these things.

**Reply by probonopd** [reply]
>  I would personally just like a "select" and "back" button

Agree. We already have the "select" button (the button built into the rotary encoder - could also wire up some other button using that same pin instead). So what we'd really need to do is add support for a secondary button, which would take the role of the "back" button. (In case the user opts into a two-button system.)

**Reply by IrregularShed** [reply]
Yep this is what I had in mind, single click on the encoder for select and a back button. It's a bit more logical I think.

I started trying to look at how to do it in Circle myself, got confused, then contracted covid again so I feel like the universe is telling me something...

**Reply by probonopd** [reply]
Get well soon!

**Reply by stevelittlefish** [reply]
> > I would personally just like a "select" and "back" button
> 
> Agree. We already have the "select" button (the button built into the rotary encoder - could also wire up some other button using that same pin instead). So what we'd really need to do is add support for a secondary button, which would take the role of the "back" button. (In case the user opts into a two-button system.)

This is exactly what I had in mind!  This way we can add the back button with only a single GPIO pin.

The other thing I was thinking is that we can add support for 4 way arrow keys by allowing the other 2 buttons to just use the encoder pins.  Then there can be a setting in the ini file which tells it to use those pins as buttons instead of encoder inputs.

This way we can support the current scheme, or an encoder with an additional back button, or 4 buttons, all with just one extra pin.

**Reply by diyelectromusic** [reply]
> The other thing I was thinking is that we can add support for 4 way arrow keys by allowing the other 2 buttons to just use the encoder pins. Then there can be a setting in the ini file which tells it to use those pins as buttons instead of encoder inputs.

Almost like we'd need an up/down/left/right/select type interface.... oh wait ;)

> This way we can support the current scheme, or an encoder with an additional back button, or 4 buttons, all with just one extra pin.

That is the whole point.  At the end of the day it is just buttons on pins tied to actions (no matter what we call them), all configured in the ini file...?  This was just a "minimum viable option" after spending several evenings hacking about with an unfamiliar codebase to get the raw structure in place for a "button user interface" to try out some ideas...

Things to think about, if this seems useful (but happy to be told not, so I don't waste my time):
* Fix the physical button logic (it's the wrong way round I realised this morning).
* Add support for the Circle GPIO manager which is the way to make it interrupt driven (I'm not convinced it is needed, polling seems ok on limited testing).
* Try some actual hardware measurements to see how sophisticated the code debouncing needs to be (again, it is the most basic that might work for now).
* Add a "two button" option - that is just the naming of items in the config file really, now the core glue code is in place.
* I might also be nice to tie these same events to MIDI...

I also have one of those shields mentioned elsewhere on order - I had that in mind from the start (I agree it would be nice to plug in a cheap, off the shelf shield and have a crude user interface straight away).

Which means there is also a future option to consider other button IO schemes.  So far I've seen the following:
* Direct GPIO (which is what I'm using here) - many cheap add-ons do this.
* I2C IO expansion (might not be worth the effort if I'm honest) - an Adafruit add-on uses this.
* Analog "voltage divider" buttons - I've seen at least one add-on that just uses A0 and resistors to "decode" 5 buttons.  I don't know if this would be worth it tbh...

Anyway, back to the original question - is there value in this or is someone else already thinking about it?

**Reply by stevelittlefish** [reply]
The interface has no concept of up, down, left and right so those names are meaningless.  If I just want a back button, which arrow direction would that be?  If you name the buttons after what they actually do it makes more sense, i.e. "menu_next" always selects the next option in the menu.

Interrupts are not *necessary* just to add this feature.  But what if we end up adding lots more button options for different things?  Maybe in the future there will be many different UI options and polling them all could cause issues.  Also it's nice to do things consistently.  These ideas aren't about just getting your feature implemented as quickly as possible, they are about creating a maintainable codebase that doesn't end up as a complete mess.

Personally for the debouncing, I would just do something simple and test it to see if it works.  If it causes issues later or on some specific setup then a more complicated scheme can be implemented.  It's better not to "fix" a problem until it actually exists.

As far as I can see, there are a total of 5 possible buttons:

- Next menu item (encoder clockwise)
- Previous menu item (encode anti-clockwise)
- Select / enter (encoder click)
- Back (encoder double-click)
- Home (encoder triple-click)

If we can get them all bound to GPIO pins using interrupts and simple debouncing I think that is all that is needed.  I don't think there is any value adding support for more complicated schemes (I2C, voltage divider etc.) as it will just clutter up the code and make it harder to configure.

**Reply by diyelectromusic** [reply]
Buttons now renamed; 2-button mode now possible; button logic and debouncing seems to work ok (for me) without interrupts.  So it would be really good if someone who knows C++ better than me could take a look and see what they think...

https://github.com/diyelectromusic/MiniDexed/tree/buttongpio

Kevin

**Reply by stevelittlefish** [reply]
@diyelectromusic It looks good to me, but I'm not familiar with circle.  It would be good if someone more familiar with the project could have a quick look.

I will try to get my raspberry pi set up with all of the hardware in the next couple of days and test it too.  I think your debounce strategy looks absolutely fine - only thing that might need tweaking is the value of DEBOUNCER.

**Reply by stevelittlefish** [reply]
@diyelectromusic I can't get your branch to compile.  I get the following error:

```
userinterface.cpp: In member function 'bool CUserInterface::Initialize()':
userinterface.cpp:73:45: error: no matching function for call to 'CHD44780Device::CHD44780Device(CI2CMaster*&, unsigned int&, const unsigned int&, const unsigned int&)'
   73 |        CConfig::LCDColumns, CConfig::LCDRows);
      |                                             ^
In file included from userinterface.h:27,
                 from userinterface.cpp:20:
../circle-stdlib/libs/circle/addon/display/hd44780device.h:69:2: note: candidate: 'CHD44780Device::CHD44780Device(unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, boolean)'
   69 |  CHD44780Device (unsigned nColumns, unsigned nRows,
      |  ^~~~~~~~~~~~~~
../circle-stdlib/libs/circle/addon/display/hd44780device.h:69:2: note:   candidate expects 10 arguments, 4 provided
../circle-stdlib/libs/circle/addon/display/hd44780device.h:55:7: note: candidate: 'constexpr CHD44780Device::CHD44780Device(const CHD44780Device&)'
   55 | class CHD44780Device : public CDevice /// LCD dot-matrix display driver (using HD44780 controller)
      |       ^~~~~~~~~~~~~~
../circle-stdlib/libs/circle/addon/display/hd44780device.h:55:7: note:   candidate expects 1 argument, 4 provided
  CC    ../CMSIS_5/CMSIS/DSP/Source/SupportFunctions/SupportFunctions.o
  CC    ../CMSIS_5/CMSIS/DSP/Source/BasicMathFunctions/BasicMathFunctions.o
  CC    ../CMSIS_5/CMSIS/DSP/Source/FastMathFunctions/FastMathFunctions.o
  CC    ../CMSIS_5/CMSIS/DSP/Source/FilteringFunctions/FilteringFunctions.o
make: *** [../circle-stdlib/libs/circle/Rules.mk:144: userinterface.o] Error 1
make: *** Waiting for unfinished jobs....
```
Can you confirm the commit hash I need to checkout for circle-stdlib and CMSIS_5?

**Reply by arsamus** [reply]
Hi. I had got the same issue. Try one of this when clone circle reposotory:

'# Get develop branch of circle
git submodule update --init --recursive
cd circle-stdlib/libs/circle
git checkout ae22928 # develop
cd -
'
Or

'# Get develop branch of circle
git submodule update --init --recursive
cd circle-stdlib/libs/circle
git checkout c9a4815 # develop
cd -
'

**Reply by stevelittlefish** [reply]
@arsamus Thanks! `git checkout c9a4815` did the trick!

**Reply by stevelittlefish** [reply]
This build doesn't work for me - the LCD screen just shows a single underscore.

@probonopd @IrregularShed Whats the verdict on polling vs interrupts?  If polling is not ever going to be merged into the codebase I can look at interrupts.  If this polling solution is ok I can try to get this working and ready to submit a merge request.

**Reply by probonopd** [reply]
I have opened a Pull Request that builds all commits to @diyelectromusic's branch so that one can easily download builds from the "Summary" section of the GitHub Actions page, e.g., https://github.com/probonopd/MiniDexed/actions/runs/2442627974.

> Whats the verdict on polling vs interrupts?

Would be good to have @rsta2's advice here.

**Reply by diyelectromusic** [reply]
> @probonopd @IrregularShed Whats the verdict on polling vs interrupts?

I don't have a major issue either way, but in general with microcontrollers, I tend to reserve interrupts for things that are either a) critical at MCU clock speeds b) real time or real low-latency; or c) absolute highest priority.  Button presses at "human speed" are none of these, so I'd err on the side of "processing when convenient" within the context of the audio synthesis rather than "stop everything, the human has pressed a button" which could even add jitter to the audio processes.

So I believe polling is more than adequate given the speed of the processor and everything else that is going on, but without knowing the full scheduling algorithm in place (i.e. what else is already polled, what the native "scan time" of the update interval is, and what else in MiniDexed is interrupt driven - I'd hope the audio output is regular off a timer interrupt for example) it is hard to tell for sure.

And this is without considering any multi-core issues - again, I am not familiar with the code enough to know how it all shares across the processor.

Kevin

**Reply by stevelittlefish** [reply]
@probonopd @diyelectromusic

I want to checkout @diyelectromusic's branch, and then merge the master branch from MiniDexed into it to try to get it to build properly and fix the screen issue.  What's the best way for me to do it?

Should I fork @diyelectromusic's project?  I want to get commits from both projects and combine them.

_Edit: I've managed to merge this branch in the main branch locally - it works fine with no conflicts once an assertion is deleted (see below)_

**Reply by stevelittlefish** [reply]
@diyelectromusic I connected the raspberry pi up to a monitor and it gives the following error message:

```
uibuttons.cpp(120): assertion failed: !m_pEventHandler
```

If I comment out the assertion everything seems to work though.  Is it safe to just delete the assertion?  Would it be easiest for you to fix it in your repository ready to be pulled into the main repo?

I've tested the select and back button and they work as expected.  I had a couple of double triggers of the back button but I think this might just be the poor quality push button I am using to test it.

I've so far only tested the UI - I haven't wired up my sound card yet but am assuming it won't affect the audio.

What do we need to do next to get this merged into the main codebase?

**Reply by probonopd** [reply]
> What do we need to do next to get this merged into the main codebase?

Please see https://github.com/probonopd/MiniDexed/pull/271.

And it needs testing in different configurations.

**Reply by diyelectromusic** [reply]
> m_pEventHandler

If that is happening then it probably means something is trying to register the event handler twice which sounds like a bug to me, although I can't see where that would be happening in the logic unless something is trying to initialise the UI twice further up.  We'd probably need a call trace to see what was calling what at the time...

Update: Perhaps it it uninitialised - I suspect it should be set to null in the CUIButtons constructor...

Kevin

**Reply by stevelittlefish** [reply]
> > m_pEventHandler
> 
> If that is happening then it probably means something is trying to register the event handler twice which sounds like a bug to me, although I can't see where that would be happening in the logic unless something is trying to initialise the UI twice further up. We'd probably need a call trace to see what was calling what at the time...
> 
> Update: Perhaps it it uninitialised - I suspect it should be set to null in the CUIButtons constructor...
> 
> Kevin

Yeah it was uninitialised.  I've fixed it in my branch.

**Comment by positionhigh** [comment]
just wanted you guys to keep updated how the things are going on the teensy front.
First PCB has arrived and is currently in testing, also a case - it looks so simple but there are a lot of issues to concider.

https://youtu.be/G6lSqf8H_4g

**Comment by sou7611** [comment]
The interesting part of this as a solution is there are display modules
that actually come with four arrow buttons on them

**Comment by JohnConnor84** [comment]
I still have some membrane keypads with GPIO conectivity to giveaway
. My original idea was to assign the pads to arrows and menu buttons.

**Reply by diyelectromusic** [reply]
> I still have some membrane keypads with GPIO conectivity to giveaway . My original idea was to assign the pads to arrows and menu buttons.

If they have a common GND and then just four or five IO pins (one per button), then they should just work with this... Do you have a link to them so I can have a look?

**Comment by JohnConnor84** [comment]
Just google Arduino 4x4 membrane keypad. If you are based in Germany or EU
I can send you one because I have 4 extra
.

**Reply by diyelectromusic** [reply]
> Arduino 4x4 membrane keypad

Ah sorry I misunderstood, you mean those common 16 button matrix type things like a phone keypad?  No this isn't supported at it requires "scanning" as rows and columns and there isn't anything in the code to do that.  It could be added of course if someone was really interested in having it.

In principle you could just use one column of course, having a single common GND pin and then four pins for each button.  What were you thinking of using it for?  Could you wire it up to an Arduino and use it as a MIDI controller for example (e.g. as a numeric keypad to select patches or something like that)?

**Comment by JohnConnor84** [comment]
Yes that was the idea. Assigning each number to OSC, LFO, EG, etc plus
arrow buttons to navigate and access menu. Maybe an Arduino nano would work
but Im not a programmer.

**Comment by dcoredump** [comment]
> Yes that was the idea. Assigning each number to OSC, LFO, EG, etc plus
> arrow buttons to navigate and access menu. Maybe an Arduino nano would work
> but Im not a programmer.

Why not adding a lvgl and touch based GUI via HDMI? Only an additional touch screen is needed (and some programming skills for creating the UI), see #62

**Comment by Kirtai** [comment]
A UI compatible with the mt-32pi hats like ClumsyMIDI would be nice. Same hardware for both systems.

**Reply by diyelectromusic** [reply]
> A UI compatible with the mt-32pi hats like ClumsyMIDI would be nice. Same hardware for both systems.

That's dependent on ssd1306 display support in Circle I'm afraid... (and Clumsy MIDI has now buttons/encoder), but when it happens, it is possible, see https://github.com/probonopd/MiniDexed/issues/169#issuecomment-1128040226

**Comment by positionhigh** [comment]
kind of out of topic, but don't know where else to post.
some new development on the teensy front:
https://www.youtube.com/watch?v=WFc2IrTSYDQ&list=PLHTypoMU1QoGOXPli8bjR6MknPiQpubHl&index=4

**Reply by probonopd** [reply]
Feel free to open a new thread in https://github.com/probonopd/MiniDexed/discussions/categories/show-and-tell if you like (even if it's about a "MiniDexed Friends" project and not MiniDexed itself).
